ada_feeding/ada_feeding/behaviors/acquisition/compute_food_frame.py

```diff
@@ -5,56 +5,246 @@
-        # Initiatilize the behavior
-        super().__init__(name=name)
-        pass
-
-    def setup(self, **kwargs) -> None:
-        pass
+        """
+        Blackboard Inputs
 
-    def update(self) -> py_trees.common.Status:
-        pass
+        Parameters
+        ----------
+        ros2_node (Node): ROS2 Node for reading/writing TFs
+        camera_info (geometry_msgs/CameraInfo): camera intrinsics matrix
+        mask (ada_feeding_msgs/Mask): food context, see Mask.msg
+        world_frame (string): ID of the TF frame to represent the food frame in
+        debug_food_frame (string): If len>0, TF frame to publish static transform
+                                   (relative to world_frame) for debugging purposes
+        """
+        # pylint: disable=unused-argument
+        # Arguments are handled generically in base class.
+        super().blackboard_inputs(
+            **{key: value for key, value in locals().items() if key != "self"}
+        )
 
-class FlipFoodFrame(py_trees.behavior.Behavior):
-    """
-    TODO
-    Take the food frame from the blackboard and 
-    rotate it PI about the world frame +Z axis.
-    """
-
-    def __init__(
+    def blackboard_outputs(
         self,
-        name: str,
-        node: Node,
-        frame_blackboard_key: str,
-        world_frame_id: str,
+        action_select_request: Optional[BlackboardKey],  # AcquisitionSelect.Request
+        food_frame: Optional[BlackboardKey],  # TransformStamped
+        debug_tf_publisher: Optional[
+            BlackboardKey
+        ] = None,  # StaticTransformBroadcaster
     ) -> None:
-        # Initiatilize the behavior
-        super().__init__(name=name)
-        pass
+        """
+        Blackboard Outputs
+        By convention (to avoid collisions), avoid non-None default arguments.
+
+        Parameters
+        ----------
+        action_select_request (AcquisitionSelect.Request): request to send to AcquisitionSelect
+                                                           (copies mask input)
+        food_frame (geometry_msgs/TransformStamped): transform from world_frame to food frame
+        debug_tf_publisher (StaticTransformBroadcaster): If set, store
+                                                        static broadcaster here to keep it alive
+                                                        for debugging purposes.
+        """
+        # pylint: disable=unused-argument
+        # Arguments are handled generically in base class.
+        super().blackboard_outputs(
+            **{key: value for key, value in locals().items() if key != "self"}
+        )
+
+    # pylint: disable=attribute-defined-outside-init
+    # It is okay for attributes in behaviors to be
+    # defined in the setup / initialise functions.
+
+    def setup(self, **kwargs):
+        """
+        Middleware (i.e. TF) setup
+        """
+
+        # Create TF Tree
+        self.tf_buffer = Buffer()
+        self.tf_listener = TransformListener(
```

Comment:
[Unregister the listener](https://github.com/ros/geometry2/blob/bfcd60628a75519b9f59864c12e742f0e24184d6/tf2_ros/src/tf2_ros/transform_listener.py#L62) in `shutdown`. Shutdown is supposed to do the inverse of `setup`. That way, after `shutdown` is called, the `tf_listener` will cease to take up resources in the callback group.

(As a side-note, we should discuss whether `create_action_servers.py` should call `setup` and `shutdown` per action execution, or once for the life of the tree. We can discuss this in-person.)

Reply:
Switching to global TF Listener / Buffer Object w/ default Node Callback group

Reply:
As discussed